presto-main/src/main/java/com/facebook/presto/resourcemanager/DistributedQueryInfoResource.java

```diff
@@ -0,0 +1,155 @@
+            @Context HttpServletRequest servletRequest,
+            @Context UriInfo uriInfo,
+            @Suspended AsyncResponse asyncResponse)
+    {
+        try {
+            Set<InternalNode> coordinators = internalNodeManager.getCoordinators();
+            ImmutableList.Builder<ListenableFuture<List<QueryStateInfo>>> queryStateInfoFutureBuilder = ImmutableList.builder();
+            for (InternalNode coordinator : coordinators) {
+                queryStateInfoFutureBuilder.add(getQueryStateFromCoordinator(servletRequest, xForwardedProto, uriInfo, coordinator));
+            }
+            List<ListenableFuture<List<QueryStateInfo>>> queryStateInfoFutureList = queryStateInfoFutureBuilder.build();
+            Futures.whenAllComplete(queryStateInfoFutureList).call(() -> {
+                try {
+                    List<QueryStateInfo> queryStateInfoList = new ArrayList<>();
+                    for (Future<List<QueryStateInfo>> queryStateInfoFuture : queryStateInfoFutureList) {
+                        queryStateInfoList.addAll(queryStateInfoFuture.get());
+                    }
+                    return asyncResponse.resume(Response.ok(queryStateInfoList).build());
+                }
+                catch (Exception ex) {
+                    log.error(ex, "Error in getting query info from one of the coordinators");
+                    return asyncResponse.resume(Response.serverError().entity(ex.getMessage()).build());
+                }
+            }, executor);
+        }
+        catch (Exception ex) {
+            log.error(ex, "Error in getting query info");
+            asyncResponse.resume(Response.serverError().entity(ex.getMessage()).build());
+        }
+    }
+
+    @GET
+    @Path("{queryId}")
+    @Produces(MediaType.APPLICATION_JSON)
+    public void getQueryStateInfo(@PathParam("queryId") QueryId queryId,
+            @HeaderParam(X_FORWARDED_PROTO) String xForwardedProto,
+            @Context UriInfo uriInfo,
+            @Context HttpServletRequest servletRequest,
+            @Suspended AsyncResponse asyncResponse)
+            throws WebApplicationException
+    {
+        proxyQueryInfoResponse(servletRequest, asyncResponse, uriInfo, queryId);
+    }
+
+    private ListenableFuture<List<QueryStateInfo>> getQueryStateFromCoordinator(HttpServletRequest servletRequest, String xForwardedProto, UriInfo uriInfo, InternalNode coordinatorNode)
+            throws IOException
+    {
+        String scheme = isNullOrEmpty(xForwardedProto) ? uriInfo.getRequestUri().getScheme() : xForwardedProto;
+        URI uri = uriInfo.getRequestUriBuilder()
+                .queryParam("includeLocalQueryOnly", true)
+                .scheme(scheme)
+                .host(coordinatorNode.getHostAndPort().toInetAddress().getHostName())
+                .port(coordinatorNode.getInternalUri().getPort())
+                .build();
+        return proxyHelper.getResponse(servletRequest, uri, jsonCodec);
```

Comment:
I think the logic here is simple enough, and not proxy-like enough, that it should just be implemented inline in this class.  i.e. let's remove the changes to add this to the RM proxy, as it's not really proxying, it's just performing the request.

Reply:
Makes sense since this class is meant to be used for proxy like stuff. I was using it like a client helper earlier.